qiskit/primitives/containers/task_result.py

```diff
@@ -0,0 +1,32 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Base Task class
+"""
+
+from __future__ import annotations
+
+from pydantic import Field
+
+from .data_bin import DataBin
+from .dataclasses import frozen_dataclass
+
+
+@frozen_dataclass
+class TaskResult:
```

Comment:
General question: What is the advantage of using dataclasses for these containers instead of just making our own regular class containers? dataclasses are note the best for subclassing because the order of attributes cant be changed or modified, and if a base has a default field, a subclass cant add any non-default fields to it. I think it would be better long term to avoid using dataclasses for the tasks and results or any other class we intend to use with subclassing or where we are adding public methods to them.

Reply:
There are pros and cons for using dataclasses.
Pros
1. Data validation (Type hints can be enforced)
2. As for options, we are supposed to use dataclasses from the RFC, so they can be common.

Cons
1. There will certainly be overhead.

I don't understand the issues related to inheritance. For example,
```python
from qiskit.primitives.containers.dataclasses import frozen_dataclass

@frozen_dataclass
class A:
    name: str = "a"

@frozen_dataclass
class B(A):
    number: int = 1
    
    def print(self):
        print(self.name, self.number)

b = B(name="b", number=3.)
b.print() # outputs b 3
```
works well. Also, the first mentioned validation works as follows:
![image](https://github.com/Qiskit/qiskit/assets/6814928/2c5e553b-669d-4b53-9e8e-50891e63c313)


@jyu00 Do you have a comment? The data structure and serialization part is important on the runtime side?

Reply:
Using a dataclass is very useful for options, especially with pydantic, since it does validation and converts dictionary inputs to nested dataclasses. It's probably less important for output that since we control that data (v.s. input data from users). 

@chriseclectic Do you have a specific use case in mind?